File: src/main/java/org/raku/structureView/RakuStructureViewElement.java
package org.raku.structureView;

import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.util.treeView.smartTree.TreeElement;
import com.intellij.navigation.ItemPresentation;
import com.intellij.psi.util.PsiTreeUtil;
import org.raku.RakuIcons;
import org.raku.extensions.RakuFrameworkCall;
import org.raku.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.*;

public class RakuStructureViewElement implements StructureViewTreeElement {
    private final RakuPsiElement element;
    private final ItemPresentation calculatedPresentation;

    public RakuStructureViewElement(RakuPsiElement element) {
        this.element = element;
        this.calculatedPresentation = getPresentation();
    }

    public RakuStructureViewElement(RakuPsiElement element, ItemPresentation calculatedPresentation) {
        this.element = element;
        this.calculatedPresentation = calculatedPresentation;
    }

    @Override
    public TreeElement @NotNull [] getChildren() {
        List<StructureViewTreeElement> structureElements = new ArrayList<>();
        if (element instanceof RakuPsiScope) {
            // Add declarations.
            List<RakuPsiDeclaration> declarations = ((RakuPsiScope)element).getDeclarations();
            Set<RakuPsiDeclaration> declSet = new HashSet<>();
            for (RakuPsiDeclaration decl : declarations) {
                if (applicable(decl)) {
                    structureElements.add(new RakuStructureViewElement(decl));
                    declSet.add(decl);
                }
            }

            // Look through statements to see if there's any DSL-like framework
            // calls.
            RakuFrameworkCall[] extensions = RakuFrameworkCall.EP_NAME.getExtensions();
            collectFrameworkCalls(structureElements, extensions, declSet, element);
        }
        return structureElements.toArray(StructureViewTreeElement.EMPTY_ARRAY);
    }

    private static void collectFrameworkCalls(List<StructureViewTreeElement> structureElements,
                                              RakuFrameworkCall[] extensions,
                                              Set<RakuPsiDeclaration> stoppers,
                                              RakuPsiElement search) {
        // If we hit a stopper, don't enter it; it's another declaration in the tree.
        if (search instanceof RakuPsiDeclaration && stoppers.contains(search))
            return;

        // See if this element is a DSL call.
        if (search instanceof RakuSubCall call) {
            for (RakuFrameworkCall ext : extensions) {
                if (ext.isApplicable(call)) {
                    ItemPresentation presentation = ext.getStructureViewPresentation(call, ext.getFrameworkData(call));
                    structureElements.add(new RakuStructureViewElement(call, presentation));
                    return;
                }
            }
        }

        // Otherwise, walk children.
        RakuPsiElement[] children = PsiTreeUtil.getChildrenOfType(search, RakuPsiElement.class);
        if (children != null)
            for (RakuPsiElement child : children)
                collectFrameworkCalls(structureElements, extensions, stoppers, child);
    }

    private static boolean applicable(RakuPsiElement child) {
        return child instanceof RakuFile ||
               child instanceof RakuPackageDecl && child.getName() != null ||
               child instanceof RakuRoutineDecl && child.getName() != null ||
               child instanceof RakuRegexDecl && child.getName() != null ||
               child instanceof RakuConstant && child.getName() != null ||
               (child instanceof RakuVariableDecl && ((RakuVariableDecl)child).getScope().equals("has")
                && child.getName() != null) ||
               child instanceof RakuSubset && child.getName() != null ||
               child instanceof RakuEnum && child.getName() != null;
    }

    @NotNull
    @Override
    public ItemPresentation getPresentation() {
        if (calculatedPresentation != null)
            return calculatedPresentation;
        if (element instanceof RakuFile) {
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    return element.getContainingFile().getName();
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Override
                public Icon getIcon(boolean b) {
                    return RakuIcons.CAMELIA;
                }
            };
        }
        if (element instanceof RakuPackageDecl)
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    RakuPackageDecl pkg = (RakuPackageDecl)element;
                    return pkg.getPackageName();
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Nullable
                @Override
                public Icon getIcon(boolean b) {
                    return RakuIcons.iconForPackageDeclarator(((RakuPackageDecl)element).getPackageKind());
                }
            };
        if (element instanceof RakuRegexDecl)
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    return ((RakuRegexDecl)element).getSignature();
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Override
                public Icon getIcon(boolean b) {
                    return RakuIcons.REGEX;
                }
            };
        if (element instanceof RakuRoutineDecl)
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    return ((RakuRoutineDecl)element).getSignature();
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Override
                public Icon getIcon(boolean b) {
                    if (((RakuRoutineDecl)element).getRoutineKind().equals("method"))
                        return RakuIcons.METHOD;
                    return RakuIcons.SUB;
                }
            };
        if (element instanceof RakuConstant)
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    return ((RakuConstant)element).getConstantName();
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Override
                public Icon getIcon(boolean b) {
                    return RakuIcons.CONSTANT;
                }
            };
        if (element instanceof RakuVariableDecl)
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    return String.join(", ", ((RakuVariableDecl)element).getVariableNames());
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Override
                public Icon getIcon(boolean b) {
                    return RakuIcons.ATTRIBUTE;
                }
            };
        if (element instanceof RakuSubset)
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    return ((RakuSubset)element).getSubsetName();
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Override
                public Icon getIcon(boolean b) {
                    return RakuIcons.SUBSET;
                }
            };
        if (element instanceof RakuEnum)
            return new ItemPresentation() {
                @Override
                public String getPresentableText() {
                    return ((RakuEnum)element).getEnumName();
                }

                @Nullable
                @Override
                public String getLocationString() {
                    return null;
                }

                @Override
                public Icon getIcon(boolean b) {
                    return RakuIcons.ENUM;
                }
            };
        throw new IllegalArgumentException();
    }

    @Override
    public RakuPsiElement getValue() {
        return element;
    }

    @Override
    public void navigate(boolean requestFocus) {
        element.navigate(requestFocus);
    }

    @Override
    public boolean canNavigate() {
        return element.canNavigate();
    }

    @Override
    public boolean canNavigateToSource() {
        return element.canNavigateToSource();
    }
}

File: src/main/java/org/raku/comma/structureView/RakuStructureViewElement.kt
package org.raku.comma.structureView

import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.util.treeView.smartTree.TreeElement
import com.intellij.navigation.ItemPresentation
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.containers.toArray
import org.raku.comma.extensions.RakuFrameworkCall
import org.raku.comma.psi.*

class RakuStructureViewElement : StructureViewTreeElement {
    private val element: RakuPsiElement
    private val calculatedPresentation: ItemPresentation?

    constructor(element: RakuPsiElement) {
        this.element = element
        this.calculatedPresentation = getPresentation()
    }

    constructor(element: RakuPsiElement, calculatedPresentation: ItemPresentation?) {
        this.element = element
        this.calculatedPresentation = calculatedPresentation
    }

    override fun getChildren(): Array<TreeElement?> {
        val structureElements: MutableList<StructureViewTreeElement?> = ArrayList<StructureViewTreeElement?>()
        if (element is RakuPsiScope) {
            // Add declarations.
            val declarations = element.getDeclarations()
            val declSet: MutableSet<RakuPsiDeclaration?> = HashSet<RakuPsiDeclaration?>()
            for (decl in declarations) {
                if (applicable(decl)) {
                    structureElements.add(RakuStructureViewElement(decl!!))
                    declSet.add(decl)
                }
            }

            // Look through statements to see if there's any DSL-like framework
            // calls.
            val extensions = RakuFrameworkCall.EP_NAME.extensions
            collectFrameworkCalls(structureElements, extensions, declSet, element)
        }
        return structureElements.toArray<TreeElement?>(StructureViewTreeElement.EMPTY_ARRAY as Array<TreeElement?>)
    }

    override fun getPresentation(): ItemPresentation {
        if (calculatedPresentation != null) return calculatedPresentation

        return when (element) {
            is RakuFile         -> rakuFileStructureElement(element)
            is RakuPackageDecl  -> rakuPackageDeclStructureElement(element)
            is RakuRegexDecl    -> rakuRegexDeclStructureElement(element)
            is RakuRoutineDecl  -> rakuRoutineDeclStructureElement(element)
            is RakuConstant     -> rakuConstantStructureElement(element)
            is RakuVariableDecl -> rakuVariableDeclStructureElement(element)
            is RakuSubset       -> rakuSubsetStructureElement(element)
            is RakuEnum         -> rakuEnumStructureElement(element)

            else                -> throw IllegalArgumentException()
        }
    }

    override fun getValue(): RakuPsiElement {
        return element
    }

    override fun navigate(requestFocus: Boolean) {
        element.navigate(requestFocus)
    }

    override fun canNavigate(): Boolean {
        return element.canNavigate()
    }

    override fun canNavigateToSource(): Boolean {
        return element.canNavigateToSource()
    }

    companion object {
        private fun collectFrameworkCalls(
            structureElements: MutableList<StructureViewTreeElement?>,
            extensions: Array<RakuFrameworkCall>,
            stoppers: MutableSet<RakuPsiDeclaration?>,
            search: RakuPsiElement?
        ) {
            // If we hit a stopper, don't enter it; it's another declaration in the tree.
            if (search is RakuPsiDeclaration && stoppers.contains(search)) return

            // See if this element is a DSL call.
            if (search is RakuSubCall) {
                for (ext in extensions) {
                    if (ext.isApplicable(search)) {
                        val presentation = ext.getStructureViewPresentation(search, ext.getFrameworkData(search))
                        structureElements.add(RakuStructureViewElement(search, presentation))
                        return
                    }
                }
            }

            // Otherwise, walk children.
            val children = PsiTreeUtil.getChildrenOfType<RakuPsiElement?>(search, RakuPsiElement::class.java)
            if (children != null) for (child in children) collectFrameworkCalls(
                structureElements,
                extensions,
                stoppers,
                child
            )
        }

        private fun applicable(child: RakuPsiElement?): Boolean {
            return child is RakuFile || child is RakuPackageDecl && child.getName() != null || child is RakuRoutineDecl && child.getName() != null || child is RakuRegexDecl && child.getName() != null || child is RakuConstant && child.getName() != null ||
                    (child is RakuVariableDecl && child.getScope() == "has"
                            && child.getName() != null) || child is RakuSubset && child.getName() != null || child is RakuEnum && child.getName() != null
        }
    }
}
